Add basic functionality

eternoendless opened the pull request:
This includes
- Decimal number creation
- Comparison
- Sign change
- Addition
- Subtraction
- Rounding

eternoendless committed:
Add basic functionality

This includes
- Decimal number creation
- Comparison
- Sign change
- Addition
- Subtraction
- Rounding

And 100% test coverage :)

eternoendless committed:
Add travis and codacy

eternoendless committed:
Use trusty distro for travis

eternoendless committed:
Drop HHVM (for now)

toutantic commented on src/Operation/Subtraction.php:
```diff
@@ -0,0 +1,189 @@
+     */
+    public function compute(DecimalNumber $a, DecimalNumber $b)
+    {
+        if (!$this->forceOwnImplementation && function_exists('bcadd')) {
```
You test if bcadd exist but then use bcsub

eternoendless replied:
Whoops, my bad

toutantic commented on README.md:
````diff
@@ -0,0 +1,154 @@
+## Testing
+
+``` bash
+$ phpunit
````
composer install
vendor/bin/phpunit

eternoendless replied:
I'll change that, thanks

toutantic commented on src/Number.php:
```diff
@@ -0,0 +1,401 @@
+     */
+    public function getPrecision()
+    {
+        return $this->exponent;
```
return getExponent()

eternoendless replied:
Sounds reasonable

eternoendless committed:
Fix test instructions on README

eternoendless committed:
Fix usage direct access to $this->exponent on getPrecision()

eternoendless committed:
Fix tests' namespaces

eternoendless committed:
Drop useOwnImplementation in favor of two separate methods (and test both directly)

LittleBigDev commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,33 @@
+
+We accept contributions via Pull Requests on [Github](https://github.com/prestashop/decimal).
+
+## Compatibility
```
PS version number missing here

LittleBigDev commented on src/Operation/Addition.php:
```diff
@@ -0,0 +1,194 @@
+    {
+        // if the addend is negative, e.g. 2 + (-1)
+        // perform subtraction instead: 2 - 1
+        if ($b->isNegative()) {
```
What about `$a` negative ? For instance : -2 + 1
Below we can see coefficients will be summed and then `$a` sign applied. This would result as -3 instead of -1

eternoendless replied:
You're absolutely right, I forgot that case. Thanks!

LittleBigDev commented on src/Number.php:
```diff
@@ -0,0 +1,401 @@
+    /**
+     * Returns the fractional part of the number.
+     * Note that this does NOT include the sign.
+     * @return string
```
Blank line missing

eternoendless replied:
I'll fix that

LittleBigDev commented on src/Number.php:
```diff
@@ -0,0 +1,401 @@
+
+    /**
+     * Returns a string representation of this object
+     * @return string
```
Blank line missing

eternoendless replied:
Will do

eternoendless committed:
Add compatibility information to CONTRIBUTING

eternoendless committed:
Fixed codacy coverage report

LittleBigDev commented on src/Operation/Rounding.php:
```diff
@@ -0,0 +1,420 @@
+     */
+    private function sanitizePrecision($precision)
+    {
+        if (!is_numeric($precision) || $precision < 0) {
```
Wouldn't a "sanitize" method have to sanitize dirty values ? Like inverting negative values ?
And only throw exception if nothing is possible ?
As is, this is a mix between validation and sanitization.

eternoendless replied:
> Like inverting negative values ?

Not necessarily (it depends on the sanitization constraints)

> As is, this is a mix between validation and sanitization.

Pretty much, yeah 🙂 

At the start, the idea was simply to make sure that precision was a valid int value (`>= 0`). At that point, I had a validation.

Then I made the design choice of making `Rounding` methods flexible enough to let them receive `'1'` instead of `1`, while still considering `'-1'` and `'true'` as invalid values. This prevented me from coercing the value to int directly because it would mean accepting any int-coercible value as a valid.

_(As a side note, the `is_numeric()` check has the side effect of accepting values like `1.12345`, `0x1`', `1e0` and `'1e0'`, but it's not a big deal)_

So, I had a validation and type coercion for certain kinds of values. I needed a single method that would allow me to consume the parameter as an int without having to worry about whether it is valid or not.

"Sanitize" was the best word I could find, but I'm always open if you find a better one 🙂

LittleBigDev replied:
Ok that makes sense :)

eternoendless committed:
Fixed unit tests not being run by travis :|

LittleBigDev commented on src/Operation/Subtraction.php:
```diff
@@ -0,0 +1,203 @@
+
+    /**
+     * Subtracts $number2 to $number1.
+     * For this algorightm to work, $number1 has to be >= $number 2.
```
Typo on word algorithm

eternoendless replied:
Thanks, I'll fix that

LittleBigDev commented on tests/NumberTest.php:
```diff
@@ -0,0 +1,497 @@
+            'NaN' => ['asd', 0],
+            'NaN with dot' => ['asd.foo', 0],
+            'NaN with point' => ['asd.foo', 0],
+            'array' => [array(), 0],
```
shouldn't it be `[[], 0]` ?

eternoendless replied:
As before, I did it on purpose 🙂

LittleBigDev commented on tests/Operation/RoundingTest.php:
```diff
@@ -0,0 +1,421 @@
+            [-1],
+            ['foo'],
+            [true],
+            [array()],
```
Shouldn't it be `[[]]` ?

eternoendless replied:
As before, I did it on purpose 🙂

LittleBigDev commented on src/Operation/Addition.php:
```diff
@@ -0,0 +1,194 @@
+            $coeff1 = str_pad($coeff1, $len2, '0', STR_PAD_LEFT);
+        }
+
+        return array($coeff1, $coeff2);
```
Shouldn't it be `[$coeff1, $coeff2]` ?

eternoendless replied:
Yeah, I'll fix that

LittleBigDev commented on src/Operation/Subtraction.php:
```diff
@@ -0,0 +1,203 @@
+            $coeff1 = str_pad($coeff1, $len2, '0', STR_PAD_LEFT);
+        }
+
+        return array($coeff1, $coeff2);
```
Shouldn't it be `[$coeff1, $coeff2]` ?

eternoendless replied:
Yeah, I'll fix that

LittleBigDev commented on tests/NumberTest.php:
```diff
@@ -0,0 +1,497 @@
+            'NaN' => ['asd'],
+            'NaN with dot' => ['asd.foo'],
+            'NaN with point' => ['asd.foo'],
+            'array' => [array()],
```
Shouldn't it be `[[]]` ?

eternoendless replied:
This time I did it on purpose to make it more evident that I'm using the array as a parameter

LittleBigDev commented on tests/NumberTest.php:
```diff
@@ -0,0 +1,497 @@
+            'empty string' => [''],
+            'NaN' => ['asd'],
+            'NaN with dot' => ['asd.foo'],
+            'NaN with point' => ['asd.foo'],
```
Maybe you wanted to provide data with a comma here ?

eternoendless replied:
That's right! Need more ☕️  😁

LittleBigDev commented on tests/Operation/AdditionTest.php:
```diff
@@ -0,0 +1,69 @@
+            // test it delegates to subtraction
+            ['1', '-2', '-1'],
+            ['-1', '-2', '-3'],
+            ['100.12345567433134123236345', '-1.1', '99.02345567433134123236345']
```
Maybe you should add `['-2', '1', '-1']` (cf. my comment above in `Addition::computeWithoutBcMath()`)

eternoendless replied:
Done!

eternoendless committed:
Fixed coverage report

eternoendless committed:
Faster composer install on travis

LittleBigDev reviewed (changes requested):
Thanks for this awesome package !
I found some little mistakes though.

eternoendless committed:
Fixed a bug in addition

Adding positive numbers to negative numbers resulted in the numbers being added as if they were both positive

eternoendless committed:
Small fixes

eternoendless committed:
Fix array syntax

eternoendless committed:
Merged ifs

LittleBigDev commented on src/Operation/Addition.php:
```diff
@@ -0,0 +1,220 @@
+            } else {
+                $result .= $sum;
+                $carryOver = 0;
+            }
```
In order to reduce cyclomatic complexity (Codacy is complaining), you could totally remove the `if` statement. (9 % 10 still makes 9).
```
$result .= $sum % 10;
$carryOver = (int) ($sum > 10);
```

eternoendless replied:
Nice one!

eternoendless committed:
Reduced cyclomatic complexity

eternoendless committed:
Expanded composer information
